src/utils/workspace-context-utils.ts

```diff
@@ -0,0 +1,119 @@
+import * as React from 'react';
+import { generatePath, useNavigate } from 'react-router-dom';
+import {
+  getActiveWorkspace,
+  k8sListResourceItems,
+  K8sResourceCommon,
+  setActiveWorkspace,
+} from '@openshift/dynamic-plugin-sdk-utils';
+import { WorkspaceModel } from '../models/workspace';
+import { useSetNamespace } from './namespace-context-utils';
+
+const workspacePathMatcher = new RegExp(/workspaces\/([^/]+)\/applications/);
+
+export type WorkspaceContextData = {
+  workspace: string;
+  workspacesLoaded: boolean;
+  workspaces: K8sResourceCommon[];
+  setWorkspace: React.Dispatch<React.SetStateAction<string>>;
+  lastUsedWorkspace: string;
+};
+
+export const WorkspaceContext = React.createContext<WorkspaceContextData>({
+  workspace: '',
+  setWorkspace: () => {},
+  workspaces: [],
+  workspacesLoaded: false,
+  lastUsedWorkspace: getActiveWorkspace(),
+});
+
+export const WorkspaceProvider = WorkspaceContext.Provider;
+
+export const useWorkspace = () => getActiveWorkspace();
+export const useLastUsedWorkspace = () => React.useContext(WorkspaceContext).lastUsedWorkspace;
+
+export const getHomeWorkspace = (workspaces) => workspaces?.find((w) => w?.status?.type === 'home');
+
+export const useActiveWorkspace = (): WorkspaceContextData => {
+  const lastUsedWorkspace = useLastUsedWorkspace();
+  const navigate = useNavigate();
+  const setActiveNamespace = useSetNamespace();
+  const [, workspaceFromUrl = ''] = window.location.pathname.match(workspacePathMatcher) || [];
+  const [workspace, setWorkspace] = React.useState<string>(getActiveWorkspace);
+  const [workspaces, setWorkspaces] = React.useState<any>([]);
+  const [workspacesLoaded, setWorkspacesLoaded] = React.useState<boolean>(false);
+
+  const getDefaultNsForWorkspace = React.useCallback((allWorkspaces, currentWorkspace) => {
+    const obj = allWorkspaces?.find((w) => w.metadata.name === currentWorkspace);
+    return obj?.status?.namespaces.find((n) => n.type === 'default');
+  }, []);
+
+  React.useEffect(() => {
+    if (workspace && workspaces?.length > 0) {
+      setActiveWorkspace(workspace);
+      const ns = getDefaultNsForWorkspace(workspaces, workspace);
+      if (ns) {
+        localStorage.setItem('activeNamespace', ns.name);
+        setActiveNamespace(ns.name);
```

Comment:
I was just wondering, would it be possible to expose the active ns through a hook? That way the ns context can use that instead of reading localStorage. Since you said the ws context is loaded first.